crashtracker/src/collector/saguard.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright 2024-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+use nix::sys::signal::{self, SaFlags, SigAction, SigHandler};
+
+// Provides a lexically-scoped guard for signals
+// During execution of the signal handler, it cannot be guaranteed that the signal is handled
+// without SA_NODEFER, thus it also cannot be guaranteed that signals like SIGCHLD and SIGPIPE will
+// _not_ be emitted during this handler as a result of the handler itself. At the same time, it
+// isn't known whether it is safe to merely block all signals, as the user's own handler will be
+// given the chance to execute after ours. Thus, we need to prevent the emission of signals we
+// might create (and cannot be created during a signal handler except by our own execution) and
+// defer any other signals.
+pub struct SaGuard<const N: usize> {
+    old_sigactions: [(signal::Signal, signal::SigAction); N],
+    old_sigmask: signal::SigSet,
+}
+
+impl<const N: usize> SaGuard<N> {
+    pub fn new(signals: &[signal::Signal; N]) -> anyhow::Result<Self> {
+        // Create an empty signal set for suppressing signals
+        let mut suppressed_signals = signal::SigSet::empty();
+        for signal in signals {
+            suppressed_signals.add(*signal);
+        }
+
+        // Save the current signal mask and block all signals except the suppressed ones
+        let mut old_sigmask = signal::SigSet::empty();
+        signal::sigprocmask(
+            signal::SigmaskHow::SIG_BLOCK,
+            Some(&suppressed_signals),
+            Some(&mut old_sigmask),
+        )?;
+
+        // Initialize array for saving old signal actions
+        // Except for SigChld and SigPipe, which are instantiated to SigIgn by default on all
```

Comment:
According to `man 7 signal` sigpipe is _not_ Ign by default (sigchld is).

Reply:
Almost all of the comments in this are wrong.  Rewriting--thank you for spotting.